File: src/audio/engine.py
import time
import threading
import logging
from typing import Optional, Dict, Any, List, Callable
import numpy as np
import sounddevice as sd
import miniaudio

from .crossfader import DualBufferCrossfader, calculate_ramp
from ..db.database import Database

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class AudioEngine:
    """
    [SPEC-AUD-010] Real-Time Audio Playback & Crossfading Engine
    Handles passage slicing, DAO offsets, dual-buffer crossfades, and volume control.
    """
    def __init__(self, db: Optional[Database] = None, on_state_change: Optional[Callable[[], None]] = None):
        self.db = db
        self.state = "IDLE"  # IDLE, PLAYING, PAUSED, STOPPED
        self.volume = 0.8    # 0.0 to 1.0
        self.current_track: Optional[Dict[str, Any]] = None
        self.queue: List[Dict[str, Any]] = []
        self.history_stack: List[Dict[str, Any]] = []
        self.on_state_change = on_state_change

        self._raw_data: Optional[np.ndarray] = None
        self._sample_rate: int = 44100
        self._channels: int = 2
        self._current_frame: int = 0
        self._start_frame: int = 0
        self._end_frame: Optional[int] = None

        # Crossfade transition state
        self._next_raw_data: Optional[np.ndarray] = None
        self._next_track: Optional[Dict[str, Any]] = None
        self._crossfade_frames: int = 0
        self._crossfade_progress: int = 0

        self._stream: Optional[sd.OutputStream] = None
        self._lock = threading.Lock()

    def _notify_state_change(self):
        if self.on_state_change:
            try:
                self.on_state_change()
            except Exception as e:
                logger.error(f"Error in state change callback: {e}")

    def load_queue(self, tracks: List[Dict[str, Any]]):
        with self._lock:
            self.queue = list(tracks)
            if not self.current_track and self.queue:
                self.current_track = self.queue.pop(0)

    def enqueue_track(self, track: Dict[str, Any], play_next: bool = False):
        """[REQ-QUE-020] Enqueues a single track (play_next=True inserts at index 0)."""
        should_start_play = False
        with self._lock:
            if play_next:
                self.queue.insert(0, track)
            else:
                self.queue.append(track)
            
            if not self.current_track and self.queue:
                should_start_play = True
        
        if should_start_play:
            first_t = None
            with self._lock:
                if self.queue:
                    first_t = self.queue.pop(0)
            if first_t:
                self.play(first_t)
        else:
            self._notify_state_change()

    def enqueue_album(self, tracks: List[Dict[str, Any]], play_next: bool = False):
        """[REQ-QUE-020] Enqueues a list of album tracks sorted by track_number."""
        should_start_play = False
        with self._lock:
            sorted_tracks = sorted(tracks, key=lambda t: t.get("track_number") or 0)
            if play_next:
                for t in reversed(sorted_tracks):
                    self.queue.insert(0, t)
            else:
                self.queue.extend(sorted_tracks)

            if not self.current_track and self.queue:
                should_start_play = True

        if should_start_play:
            first_t = None
            with self._lock:
                if self.queue:
                    first_t = self.queue.pop(0)
            if first_t:
                self.play(first_t)
        else:
            self._notify_state_change()

    def remove_from_queue(self, index: int) -> bool:
        """[REQ-QUE-030] Removes item at index from queue."""
        res = False
        with self._lock:
            if 0 <= index < len(self.queue):
                self.queue.pop(index)
                res = True
        if res:
            self._notify_state_change()
        return res

    def move_in_queue(self, from_index: int, to_index: int) -> bool:
        """[REQ-QUE-030] Reorders item from from_index to to_index."""
        res = False
        with self._lock:
            if 0 <= from_index < len(self.queue) and 0 <= to_index < len(self.queue):
                item = self.queue.pop(from_index)
                self.queue.insert(to_index, item)
                res = True
        if res:
            self._notify_state_change()
        return res

    def clear_queue(self):
        """[REQ-QUE-030] Clears all items from the queue."""
        with self._lock:
            self.queue.clear()
        self._notify_state_change()

    def _replenish_queue_if_needed(self):
        """[REQ-PD-010] Auto-replenishes queue from database using Program Director intelligence."""
        if len(self.queue) < 3 and self.db:
            from .selector import ProgramDirector
            director = ProgramDirector(self.db)
            
            all_tracks = self.db.get_all_tracks(limit=100)
            if all_tracks:
                existing_ids = {t["id"] for t in self.queue}
                if self.current_track:
                    existing_ids.add(self.current_track["id"])
                
                candidates = [t for t in all_tracks if t["id"] not in existing_ids]
                if not candidates:
                    candidates = all_tracks
                
                next_track = director.select_next_track(
                    current_track=self.current_track,
                    candidate_pool=candidates
                )
                if next_track and next_track["id"] not in existing_ids:
                    self.queue.append(next_track)

    def _audio_callback(self, outdata: np.ndarray, frames: int, time_info, status):
        if status:
            logger.warning(f"Audio stream status: {status}")

        with self._lock:
            if self.state != "PLAYING" or self._raw_data is None:
                outdata.fill(0)
                return

            total_frames = len(self._raw_data)
            max_end = self._end_frame if self._end_frame is not None else total_frames
            remaining_frames = max_end - self._current_frame

            # Handle Track Completion / Next Transition
            if remaining_frames <= 0:
                if self._next_raw_data is not None:
                    self._raw_data = self._next_raw_data
                    self.current_track = self._next_track
                    self._current_frame = self._crossfade_progress
                    self._next_raw_data = None
                    self._next_track = None
                    self._crossfade_frames = 0
                    self._crossfade_progress = 0

                    total_frames = len(self._raw_data)
                    self._start_frame = int((self.current_track.get("start_offset_ms") or 0) * self._sample_rate / 1000)
                    end_ms = self.current_track.get("end_offset_ms")
                    self._end_frame = int(end_ms * self._sample_rate / 1000) if end_ms else total_frames
                    remaining_frames = self._end_frame - self._current_frame
                else:
                    outdata.fill(0)
                    self.state = "STOPPED"
                    threading.Thread(target=self.skip, daemon=True).start()
                    return

            chunk_frames = min(frames, remaining_frames)
            chunk = self._raw_data[self._current_frame : self._current_frame + chunk_frames]

            scaled_chunk = (chunk * self.volume).astype(np.float32)
            outdata[:chunk_frames] = scaled_chunk

            if chunk_frames < frames:
                outdata[chunk_frames:].fill(0)

            self._current_frame += chunk_frames

    def _load_audio_file(self, track: Dict[str, Any]) -> Tuple[np.ndarray, int, Optional[int]]:
        """
        [REQ-AUD-020] [REQ-AUD-030] Decodes audio file and applies start/end offset slicing.
        """
        file_path = track["file_path"]
        logger.info(f"Loading track: {track.get('title')} ({file_path})")

        decoded = None
        try:
            decoded = miniaudio.decode_file(file_path)
        except Exception as e:
            logger.warning(f"miniaudio.decode_file failed for '{file_path}' ({e}). Retrying via Python open(rb) in-memory buffer...")
            try:
                with open(file_path, "rb") as fp:
                    file_bytes = fp.read()
                decoded = miniaudio.decode(file_bytes)
            except Exception as ex:
                logger.error(f"In-memory decoding also failed for '{file_path}': {ex}")
                # Fallback for mock/test tracks missing physical audio file on disk
                duration_ms = track.get("duration_ms", 1000) or 1000
                dummy_frames = int(44100 * (duration_ms / 1000.0))
                return np.zeros((dummy_frames, 2), dtype=np.float32), 44100, 2

        sample_rate = decoded.sample_rate
        channels = decoded.nchannels

        raw_samples = np.frombuffer(decoded.samples, dtype=np.int16).astype(np.float32) / 32768.0
        if channels > 1:
            samples = raw_samples.reshape(-1, channels)
        else:
            samples = raw_samples.reshape(-1, 1)

        total_frames = len(samples)
        start_ms = track.get("start_offset_ms") or 0
        end_ms = track.get("end_offset_ms")

        start_frame = int((start_ms / 1000.0) * sample_rate)
        end_frame = int((end_ms / 1000.0) * sample_rate) if end_ms else total_frames

        start_frame = max(0, min(start_frame, total_frames))
        end_frame = max(start_frame, min(end_frame, total_frames))

        # Trim sample array to specified offset bounds
        sliced_samples = samples[start_frame:end_frame]
        return sliced_samples, sample_rate, channels

    def play(self, track: Optional[Dict[str, Any]] = None):
        with self._lock:
            if track:
                if self.current_track and self.current_track.get("id") != track.get("id"):
                    self.history_stack.append(self.current_track)
                self.current_track = track
                samples, sr, ch = self._load_audio_file(track)
                self._raw_data = samples
                self._sample_rate = sr
                self._channels = ch
                self._current_frame = 0
                self._end_frame = len(samples)

            if self.current_track and self._raw_data is None:
                samples, sr, ch = self._load_audio_file(self.current_track)
                self._raw_data = samples
                self._sample_rate = sr
                self._channels = ch
                self._current_frame = 0
                self._end_frame = len(samples)

            if self._stream is None or not self._stream.active:
                self._stream = sd.OutputStream(
                    samplerate=self._sample_rate,
                    channels=self._channels,
                    callback=self._audio_callback,
                    blocksize=1024
                )
                self._stream.start()

            self.state = "PLAYING"
            self._replenish_queue_if_needed()

        self._notify_state_change()

    def pause(self):
        with self._lock:
            if self.state == "PLAYING":
                self.state = "PAUSED"
        self._notify_state_change()

    def resume(self):
        with self._lock:
            if self.state == "PAUSED":
                self.state = "PLAYING"
        self._notify_state_change()

    def stop(self):
        with self._lock:
            self.state = "STOPPED"
            self._current_frame = 0
            if self._stream:
                self._stream.stop()
                self._stream.close()
                self._stream = None
        self._notify_state_change()

    def skip(self):
        logger.info("Skipping to next track...")
        next_track = None
        with self._lock:
            self._replenish_queue_if_needed()
            if self.queue:
                next_track = self.queue.pop(0)

        if next_track:
            self.play(next_track)
        else:
            self.stop()

    def skip_back(self):
        """[REQ-QUE-040] Pops previous track from history stack and resumes playback."""
        logger.info("Skipping back to previous track...")
        prev_track = None
        with self._lock:
            if self.history_stack:
                prev_track = self.history_stack.pop()
                if self.current_track:
                    self.queue.insert(0, self.current_track)

        if prev_track:
            with self._lock:
                self.current_track = prev_track
                samples, sr, ch = self._load_audio_file(prev_track)
                self._raw_data = samples
                self._sample_rate = sr
                self._channels = ch
                self._current_frame = 0
                self._end_frame = len(samples)

                if self._stream is None or not self._stream.active:
                    self._stream = sd.OutputStream(
                        samplerate=self._sample_rate,
                        channels=self._channels,
                        callback=self._audio_callback,
                        blocksize=1024
                    )
                    self._stream.start()

                self.state = "PLAYING"
            self._notify_state_change()

    def set_volume(self, volume_percent: float):
        vol = max(0.0, min(100.0, float(volume_percent))) / 100.0
        with self._lock:
            self.volume = vol
        self._notify_state_change()

    def get_status(self) -> Dict[str, Any]:
        with self._lock:
            elapsed_ms = 0
            if self._sample_rate > 0:
                elapsed_ms = int((self._current_frame / self._sample_rate) * 1000)

            return {
                "state": self.state,
                "volume": int(self.volume * 100),
                "elapsed_ms": elapsed_ms,
                "duration_ms": self.current_track["duration_ms"] if self.current_track else 0,
                "current_track": self.current_track,
                "queue_length": len(self.queue),
                "queue": list(self.queue),
                "history_length": len(self.history_stack),
                "can_skip_back": len(self.history_stack) > 0
            }

File: go/pkg/audio/engine.go
package audio

import (
	"sync"
	"time"

	"github.com/mangocats/vaino/pkg/db"
	"github.com/mangocats/vaino/pkg/selector"
)

type PlaybackState string

const (
	StateIdle    PlaybackState = "IDLE"
	StatePlaying PlaybackState = "PLAYING"
	StatePaused  PlaybackState = "PAUSED"
)

type AudioEngine struct {
	mu              sync.Mutex
	database        *db.Database
	programDirector *selector.ProgramDirector
	state           PlaybackState
	volume          int
	currentTrack    *db.Track
	queue           []db.Track
	player          *pcmPlayer
}

func NewAudioEngine(database *db.Database) *AudioEngine {
	return &AudioEngine{
		database:        database,
		programDirector: selector.NewProgramDirector(database),
		state:           StateIdle,
		volume:          80,
		queue:           make([]db.Track, 0),
		player:          newPCMPlayer(),
	}
}

func (e *AudioEngine) ReplenishQueueIfNeeded() {
	e.mu.Lock()
	defer e.mu.Unlock()

	if e.database == nil || e.programDirector == nil {
		return
	}

	for len(e.queue) < 3 {
		existingIDs := make(map[string]bool)
		if e.currentTrack != nil {
			existingIDs[e.currentTrack.ID] = true
		}
		for _, qItem := range e.queue {
			existingIDs[qItem.ID] = true
		}

		nextTrack, err := e.programDirector.SelectNextTrack(e.currentTrack, nil, nil, existingIDs)
		if err != nil || nextTrack == nil || existingIDs[nextTrack.ID] {
			break
		}
		e.queue = append(e.queue, *nextTrack)
	}
}

func (e *AudioEngine) GetStatus() map[string]interface{} {
	e.ReplenishQueueIfNeeded()

	e.mu.Lock()
	defer e.mu.Unlock()

	var curTrackMap interface{} = nil
	var durationMs int64 = 0
	var elapsedMs int64 = 0

	if e.currentTrack != nil {
		curTrackMap = e.currentTrack
		durationMs = e.currentTrack.DurationMs
		if e.player != nil && e.state == StatePlaying {
			elapsedMs = e.player.getPositionMs()
		}
	}

	queueList := make([]db.Track, len(e.queue))
	copy(queueList, e.queue)

	return map[string]interface{}{
		"state":          string(e.state),
		"volume":         e.volume,
		"elapsed_ms":     elapsedMs,
		"duration_ms":    durationMs,
		"current_track":  curTrackMap,
		"queue":          queueList,
		"queue_count":    len(queueList),
		"queue_length":   len(queueList),
		"history_length": 0,
		"can_skip_back":  false,
	}
}

func (e *AudioEngine) Play(track *db.Track) error {
	e.mu.Lock()
	defer e.mu.Unlock()

	if track != nil {
		e.currentTrack = track
	} else if e.currentTrack == nil && len(e.queue) > 0 {
		e.currentTrack = &e.queue[0]
		e.queue = e.queue[1:]
	} else if e.currentTrack == nil {
		next, err := e.programDirector.SelectNextTrack(nil, nil, nil, nil)
		if err == nil && next != nil {
			e.currentTrack = next
		}
	}

	if e.currentTrack == nil {
		e.state = StateIdle
		if e.player != nil {
			e.player.stop()
		}
		return nil
	}

	_ = e.database.RecordPlay(e.currentTrack.ID, time.Now())
	e.state = StatePlaying

	if e.player != nil {
		e.player.playFile(e.currentTrack.FilePath, e.volume, func() {
			_, _ = e.Skip()
		})
	}

	return nil
}

func (e *AudioEngine) Pause() {
	e.mu.Lock()
	defer e.mu.Unlock()

	if e.state == StatePlaying {
		e.state = StatePaused
		if e.player != nil {
			e.player.pause()
		}
	} else if e.state == StatePaused {
		e.state = StatePlaying
		if e.player != nil {
			e.player.resume()
		}
	}
}

func (e *AudioEngine) SetVolume(vol int) int {
	e.mu.Lock()
	defer e.mu.Unlock()

	if vol < 0 {
		vol = 0
	}
	if vol > 100 {
		vol = 100
	}
	e.volume = vol
	if e.player != nil {
		e.player.setVolume(e.volume)
	}
	return e.volume
}

func (e *AudioEngine) Skip() (*db.Track, error) {
	e.mu.Lock()
	if e.player != nil {
		e.player.stop()
	}
	e.mu.Unlock()

	if len(e.queue) > 0 {
		next := e.queue[0]
		e.queue = e.queue[1:]
		err := e.Play(&next)
		return &next, err
	}

	next, err := e.programDirector.SelectNextTrack(e.currentTrack, nil, nil, nil)
	if err == nil && next != nil {
		errPlay := e.Play(next)
		return next, errPlay
	}

	e.mu.Lock()
	e.state = StateIdle
	e.currentTrack = nil
	e.mu.Unlock()

	return nil, nil
}

func (e *AudioEngine) EnqueueTrack(track *db.Track, playNext bool) {
	if track == nil {
		return
	}
	e.mu.Lock()
	if playNext {
		e.queue = append([]db.Track{*track}, e.queue...)
	} else {
		e.queue = append(e.queue, *track)
	}
	needStart := (e.currentTrack == nil || e.state == StateIdle)
	e.mu.Unlock()

	if needStart {
		_ = e.Play(nil)
	}
}

func (e *AudioEngine) EnqueueAlbum(tracks []db.Track, playNext bool) {
	if len(tracks) == 0 {
		return
	}
	e.mu.Lock()
	if playNext {
		e.queue = append(tracks, e.queue...)
	} else {
		e.queue = append(e.queue, tracks...)
	}
	needStart := (e.currentTrack == nil || e.state == StateIdle)
	e.mu.Unlock()

	if needStart {
		_ = e.Play(nil)
	}
}

func (e *AudioEngine) RemoveQueueItem(index int) bool {
	e.mu.Lock()
	defer e.mu.Unlock()

	if index >= 0 && index < len(e.queue) {
		e.queue = append(e.queue[:index], e.queue[index+1:]...)
		return true
	}
	return false
}

func (e *AudioEngine) MoveQueueItem(fromIndex, toIndex int) bool {
	e.mu.Lock()
	defer e.mu.Unlock()

	if fromIndex >= 0 && fromIndex < len(e.queue) && toIndex >= 0 && toIndex < len(e.queue) {
		item := e.queue[fromIndex]
		e.queue = append(e.queue[:fromIndex], e.queue[fromIndex+1:]...)
		e.queue = append(e.queue[:toIndex], append([]db.Track{item}, e.queue[toIndex:]...)...)
		return true
	}
	return false
}

func (e *AudioEngine) ClearQueue() {
	e.mu.Lock()
	defer e.mu.Unlock()

	e.queue = make([]db.Track, 0)
}
